coord/coordinator.go

```diff
@@ -15,30 +15,44 @@ import (
 	"github.com/plprobelab/go-kademlia/network/address"
 	"github.com/plprobelab/go-kademlia/network/endpoint"
 	"github.com/plprobelab/go-kademlia/query"
+	"github.com/plprobelab/go-kademlia/routing"
 	"github.com/plprobelab/go-kademlia/util"
 )
 
+// A StateMachine progresses through a set of states in response to transition events.
+type StateMachine[S any, E event.Action] interface {
+	// Enqueue enqueues an event to be processed by the state machine.
+	Enqueue(context.Context, E)
+	// Advance advances the state of the state machine.
+	Advance(context.Context) S
+}
```

Comment:
I think a `StateMachine` interface makes sense. I also think it's great that we can constrain the type of events that we pass into the state machine with the generic type parameters!

I'm not sure about the part where we require each state to maintain its own queue of events it needs to process. I would have expected that this is the responsibility of another component. What drove the requirement to have a separate queue in each state?

Reply:
Okay, I just read the PR description.